File: ffuzzy/wasm/test/api_parity.test.mjs
// API surface parity test — JS side.
// Verifies that the JS package exports and instance methods match the Dart
// source of truth. Any method added or removed in Dart should be reflected
// here and in the TS corpus implementation.
import assert from 'node:assert/strict';
import { test, before } from 'node:test';

import * as Mod from '../dist/ffuzzy.mjs';
const {
  ffuzzyInitialize, ffuzzyReady, FuzzyCorpus, FuzzyKey, FuzzyOptions,
  FuzzyCase, FuzzyNorm, FuzzyMode, FuzzyScoring, FuzzyKeyKind,
  fuzzyCodepointToUtf16, highlightHtml,
} = Mod;

before(async () => { await ffuzzyInitialize(); });

// ── Module-level exports ──────────────────────────────────────────────────────

test('exports: required functions and classes', () => {
  const required = [
    'ffuzzyInitialize', 'ffuzzyReady',
    'FuzzyCorpus', 'FuzzyKey', 'FuzzyOptions',
    'FuzzyCase', 'FuzzyNorm', 'FuzzyMode', 'FuzzyScoring', 'FuzzyKeyKind',
    'fuzzyCodepointToUtf16', 'highlightHtml',
  ];
  for (const name of required) {
    assert.ok(name in Mod, `missing export: ${name}`);
  }
});

test('exports: enum values match Dart', () => {
  assert.equal(FuzzyCase.respect, 0);
  assert.equal(FuzzyCase.ignore,  1);
  assert.equal(FuzzyCase.smart,   2);

  assert.equal(FuzzyNorm.never, 0);
  assert.equal(FuzzyNorm.smart, 1);

  assert.equal(FuzzyScoring.fast,   0);
  assert.equal(FuzzyScoring.off,    1);
  assert.equal(FuzzyScoring.nucleo, 2);

  assert.equal(FuzzyKeyKind.original, 0);
  assert.equal(FuzzyKeyKind.pinyin,   1);
  assert.equal(FuzzyKeyKind.initials, 2);
  assert.equal(FuzzyKeyKind.romaji,   3);
  assert.equal(FuzzyKeyKind.custom,   100);
});

test('exports: FuzzyCorpus static factories', () => {
  assert.equal(typeof FuzzyCorpus.strings, 'function', 'FuzzyCorpus.strings');
  assert.equal(typeof FuzzyCorpus.byKey,   'function', 'FuzzyCorpus.byKey');
  assert.equal(typeof FuzzyCorpus.byKeys,  'function', 'FuzzyCorpus.byKeys');
});

// ── Instance methods ──────────────────────────────────────────────────────────

test('instance: all required methods present', () => {
  const c = FuzzyCorpus.strings(['a']);

  // Search methods — must match Dart FuzzyCorpus API
  const searchMethods = [
    'fuzzy',        'fuzzyRaws',
    'substring',    'substringRaws',
    'prefix',       'prefixRaws',
    'postfix',      'postfixRaws',
    'suffix',       'suffixRaws',
    'exact',        'exactRaws',
  ];
  for (const m of searchMethods) {
    assert.equal(typeof c[m], 'function', `missing method: ${m}`);
  }

  // Mutation methods
  const mutationMethods = [
    'add', 'addAll', 'addKey',
    'update', 'removeAt', 'removeWhere',
    'refresh', 'clear',
  ];
  for (const m of mutationMethods) {
    assert.equal(typeof c[m], 'function', `missing method: ${m}`);
  }

  // Lifecycle
  assert.equal(typeof c.dispose, 'function', 'dispose');
  assert.equal(typeof c.length,  'number',   'length getter');

  c.dispose();
});

// ── FuzzyHit shape ────────────────────────────────────────────────────────────

test('FuzzyHit: all fields present (mirrors Dart FuzzyHit)', () => {
  const c = FuzzyCorpus.strings(['src/main.dart']);
  const [hit] = c.fuzzy('main', { highlight: true });

  assert.ok(hit !== undefined, 'expected a hit');

  const requiredFields = [
    'raw',            // T
    'index',          // int
    'score',          // int
    'matchedKind',    // FuzzyKeyKind (as int in JS)
    'matchedKindCode', // int — matches Dart FuzzyHit.matchedKindCode
    'matchedKey',     // int
    'indices',        // number[]
  ];
  for (const f of requiredFields) {
    assert.ok(f in hit, `FuzzyHit missing field: ${f}`);
  }

  // Type checks
  assert.equal(typeof hit.raw,             'string',  'raw: string');
  assert.equal(typeof hit.index,           'number',  'index: number');
  assert.equal(typeof hit.score,           'number',  'score: number');
  assert.equal(typeof hit.matchedKind,     'number',  'matchedKind: number');
  assert.equal(typeof hit.matchedKindCode, 'number',  'matchedKindCode: number');
  assert.equal(typeof hit.matchedKey,      'number',  'matchedKey: number');
  assert.ok(Array.isArray(hit.indices),              'indices: Array');

  // matchedKind and matchedKindCode must be equal in JS
  assert.equal(hit.matchedKind, hit.matchedKindCode, 'matchedKind === matchedKindCode');

  c.dispose();
});

// ── FuzzyKey ──────────────────────────────────────────────────────────────────

test('FuzzyKey: constructor and static factory', () => {
  const k1 = new FuzzyKey('zhangsan', 1);
  assert.equal(k1.text, 'zhangsan');
  assert.equal(k1.kind, 1);

  const k2 = FuzzyKey.kind('zs', FuzzyKeyKind.initials);
  assert.equal(k2.text, 'zs');
  assert.equal(k2.kind, FuzzyKeyKind.initials);
});

// ── Utility functions ─────────────────────────────────────────────────────────

test('fuzzyCodepointToUtf16: BMP no-op, astral doubles', () => {
  assert.deepEqual(fuzzyCodepointToUtf16('abc', [0, 2]), [0, 2]);
  // emoji '😀' occupies 2 UTF-16 code units; offset[1] should be 2
  const offsets = fuzzyCodepointToUtf16('a😀b', [0, 1, 2]);
  assert.equal(offsets[0], 0);
  assert.equal(offsets[1], 1);  // 😀 starts at UTF-16 offset 1
  assert.equal(offsets[2], 3);  // 'b' starts at UTF-16 offset 3 (after 2-unit emoji)
});

test('highlightHtml: wraps matched chars, merges adjacent, escapes HTML', () => {
  assert.equal(highlightHtml('src/main.dart', [0, 1, 2]), '<mark>src</mark>/main.dart');
  assert.ok(highlightHtml('<b>item</b>', [0]).includes('&lt;'));
  assert.equal(highlightHtml('abc', [], { tag: 'em' }), 'abc');
  assert.equal(highlightHtml('abc', [1], { tag: 'b' }), 'a<b>b</b>c');
});

File: ffuzzy/test/api_parity_test.dart
// API surface parity test — Dart side (source of truth).
// Verifies that every public method of FuzzyCorpus<T> can be called without
// error with valid inputs, and that return types carry the expected fields.
// Any method added here must also appear in wasm/test/api_parity.test.mjs.
//
//   flutter test test/api_parity_test.dart
import 'dart:io';

import 'package:flutter_test/flutter_test.dart';
import 'package:ffuzzy/ffuzzy.dart';

String _libPath() {
  final root = Directory.current.path;
  if (Platform.isWindows) return '$root\\libffz.dll';
  if (Platform.isMacOS)   return '$root/libffz.dylib';
  return '$root/build_x86_64/libffuzzy.so';
}

void main() {
  late String lib;
  setUpAll(() { lib = _libPath(); });

  // ── Static constructors ──────────────────────────────────────────────────────

  group('FuzzyCorpus static constructors', () {
    test('strings', () {
      final c = FuzzyCorpus.strings(['a', 'b'], libraryPath: lib);
      expect(c.length, 2);
      c.dispose();
    });

    test('byKey', () {
      final c = FuzzyCorpus.byKey(
        [{'name': 'Alice'}, {'name': 'Bob'}], 'name', libraryPath: lib,
      );
      expect(c.length, 2);
      c.dispose();
    });

    test('byKeys', () {
      final c = FuzzyCorpus.byKeys(
        [{'name': 'Alice', 'city': 'Boston'}], ['name', 'city'], libraryPath: lib,
      );
      expect(c.length, 1);
      c.dispose();
    });

    test('buildAsync', () async {
      final c = await FuzzyCorpus.buildAsync(
        ['alpha', 'beta'], stringOf: (s) => s, libraryPath: lib,
      );
      expect(c.length, 2);
      c.dispose();
    });
  });

  // ── Search methods ───────────────────────────────────────────────────────────

  group('FuzzyCorpus search methods', () {
    late FuzzyCorpus<String> c;
    setUp(() => c = FuzzyCorpus.strings(
      ['src/main.dart', 'lib/widget.dart', 'README.md'], libraryPath: lib,
    ));
    tearDown(() => c.dispose());

    test('fuzzy',         () { final h = c.fuzzy('main');        expect(h, isNotEmpty); });
    test('fuzzyAsync',    () async { final h = await c.fuzzyAsync('main'); expect(h, isNotEmpty); });
    test('fuzzyRaws',     () { final r = c.fuzzyRaws('main');     expect(r, isNotEmpty); });
    test('fuzzyRawsAsync',() async { final r = await c.fuzzyRawsAsync('main'); expect(r, isNotEmpty); });

    test('prefix',         () { expect(c.prefix('src'),      isNotEmpty); });
    test('prefixAsync',    () async { expect(await c.prefixAsync('src'), isNotEmpty); });
    test('prefixRaws',     () { expect(c.prefixRaws('src'),   isNotEmpty); });
    test('prefixRawsAsync',() async { expect(await c.prefixRawsAsync('src'), isNotEmpty); });

    test('postfix',         () { expect(c.postfix('.dart'),    isNotEmpty); });
    test('postfixAsync',    () async { expect(await c.postfixAsync('.dart'), isNotEmpty); });
    test('postfixRaws',     () { expect(c.postfixRaws('.dart'), isNotEmpty); });
    test('postfixRawsAsync',() async { expect(await c.postfixRawsAsync('.dart'), isNotEmpty); });

    test('suffix',          () { expect(c.suffix('.dart'),     isNotEmpty); });
    test('suffixAsync',     () async { expect(await c.suffixAsync('.dart'), isNotEmpty); });
    test('suffixRaws',      () { expect(c.suffixRaws('.dart'),  isNotEmpty); });
    test('suffixRawsAsync', () async { expect(await c.suffixRawsAsync('.dart'), isNotEmpty); });

    test('exact',          () { expect(c.exact('README.md'),   isNotEmpty); });
    test('exactAsync',     () async { expect(await c.exactAsync('README.md'), isNotEmpty); });
    test('exactRaws',      () { expect(c.exactRaws('README.md'), isNotEmpty); });
    test('exactRawsAsync', () async { expect(await c.exactRawsAsync('README.md'), isNotEmpty); });

    test('substring',         () { expect(c.substring('widget'),    isNotEmpty); });
    test('substringAsync',    () async { expect(await c.substringAsync('widget'), isNotEmpty); });
    test('substringRaws',     () { expect(c.substringRaws('widget'),  isNotEmpty); });
    test('substringRawsAsync',() async { expect(await c.substringRawsAsync('widget'), isNotEmpty); });
  });

  // ── FuzzyHit fields ──────────────────────────────────────────────────────────

  group('FuzzyHit fields', () {
    late FuzzyCorpus<String> c;
    setUp(() => c = FuzzyCorpus.strings(['src/main.dart'], libraryPath: lib));
    tearDown(() => c.dispose());

    test('all fields present', () {
      final hits = c.fuzzy('main', highlight: true);
      expect(hits, isNotEmpty);
      final h = hits.first;
      // These field names must match what JS exports in FuzzyHit
      expect(h.raw,             isA<String>());
      expect(h.index,           isA<int>());
      expect(h.score,           isA<int>());
      expect(h.matchedKind,     isA<FuzzyKeyKind>());
      expect(h.matchedKindCode, isA<int>());
      expect(h.matchedKey,      isA<int>());
      expect(h.indices,         isA<List<int>>());
      expect(h.indices, isNotEmpty); // highlight: true
    });

    test('matchedKindCode equals matchedKind.code', () {
      final h = c.fuzzy('main').first;
      expect(h.matchedKindCode, equals(h.matchedKind.code));
    });
  });

  // ── Mutation methods ─────────────────────────────────────────────────────────

  group('FuzzyCorpus mutation methods', () {
    late FuzzyCorpus<String> c;
    setUp(() => c = FuzzyCorpus.strings(['alpha', 'beta', 'gamma'], libraryPath: lib));
    tearDown(() => c.dispose());

    test('add',         () { c.add('delta'); expect(c.length, 4); });
    test('addAll',      () { c.addAll(['d', 'e']); expect(c.length, 5); });
    test('addKey',      () { c.addKey('zeta', [FuzzyKey.kind('z', FuzzyKeyKind.custom)]); expect(c.length, 4); });
    test('update',      () { c.update(0, 'ALPHA'); expect(c.fuzzy('ALPHA'), isNotEmpty); });
    test('removeAt',    () { c.removeAt(0); expect(c.length, 2); });
    test('removeWhere', () { final n = c.removeWhere((s) => s.startsWith('a')); expect(n, greaterThan(0)); });
    test('refresh',     () { c.refresh(['x', 'y']); expect(c.length, 2); });
    test('clear',       () { c.clear(); expect(c.length, 0); });
    test('addAllAsync', () async { await c.addAllAsync(['p', 'q']); expect(c.length, 5); });
  });

  // ── Lifecycle ────────────────────────────────────────────────────────────────

  group('FuzzyCorpus lifecycle', () {
    test('dispose is idempotent', () {
      final c = FuzzyCorpus.strings(['x'], libraryPath: lib);
      c.dispose(); c.dispose(); // should not throw
    });
    test('use after dispose throws', () {
      final c = FuzzyCorpus.strings(['x'], libraryPath: lib);
      c.dispose();
      expect(() => c.fuzzy('x'), throwsStateError);
    });
    test('disposeAndWait', () async {
      final c = FuzzyCorpus.strings(['x'], libraryPath: lib);
      await c.disposeAndWait();
    });
  });

  // ── Utility functions ─────────────────────────────────────────────────────────

  group('Utility functions', () {
    test('fuzzyCodepointToUtf16: BMP no-op', () {
      expect(fuzzyCodepointToUtf16('abc', [0, 2]), equals([0, 2]));
    });
    test('fuzzyCodepointToUtf16: astral doubles', () {
      // 'a😀b' → codepoints [a, 😀, b], UTF-16 offsets [0, 1, 3]
      final offsets = fuzzyCodepointToUtf16('a😀b', [0, 1, 2]);
      expect(offsets, equals([0, 1, 3]));
    });
  });

  // ── Enums ────────────────────────────────────────────────────────────────────

  group('Enum values match JS side', () {
    test('FuzzyCase codes', () {
      expect(FuzzyCase.respect.index, 0); // code = 0
      expect(FuzzyCase.ignore.index,  1);
      expect(FuzzyCase.smart.index,   2);
    });
    test('FuzzyKeyKind codes', () {
      expect(FuzzyKeyKind.original.code, 0);
      expect(FuzzyKeyKind.pinyin.code,   1);
      expect(FuzzyKeyKind.initials.code, 2);
      expect(FuzzyKeyKind.romaji.code,   3);
      expect(FuzzyKeyKind.custom.code,   100);
    });
    test('FuzzyScoring codes', () {
      expect(FuzzyScoring.fast.index,   0);
      expect(FuzzyScoring.off.index,    1);
      expect(FuzzyScoring.nucleo.index, 2);
    });
  });
}
